Core/clim-basic/clipboard.lisp

```diff
@@ -0,0 +1,166 @@
+content of the selection or the clipboard is no longer available for
+pasting.
+
+PORT is the port that the sheet belongs to. It should be the primary
+argument that individual sheets specialise on.
+
+SHEET is the owner of the selection or clipboard.
+
+Clear the clipboard if CLIPBOARD-P is true, otherwise the selection.")
+  (:method ((port clim:port) (sheet clim:sheet))
+    nil))
+
+(defgeneric local-selection-content (port)
+  (:documentation "Returns the content of the selection in this Lisp image.
+If the global selection is currently held, the value returned is the
+same as what would be sent in response to a REQUEST-CLIPBOARD-CONTENT
+call.
+
+If content is available, this function should return two values: The
+content of the selection, and the presentation-type of the object.
+Otherwise (VALUES NIL NIL) should be returned.
+
+This function is used by the copy selection to clipboard operation.
+This is in order to present the same beahviour as most non-CLIM
+applications exhibit: They are able to perform the copy
+operation (usually invoked by control-c) even when not owning the
+system-wide selection.
+
+The simplest implementation of this function would simply return the
+content of the selection, or (VALUES NIL NIL) if it is not held.")
+  (:method ((port clim:port))
+    nil))
+
+;;;
+;;;  The following functions implement the standard API to request
+;;;  content from the clipboard.
+;;;
+
+(defgeneric request-selection-content (port sheet type)
+  (:documentation "Request the content of the selection.
+
+Please refer to the documentation for REQUEST-CLIPBOARD-CONTENT for
+details.")
+  (:method ((port clim:port) (sheet clim:sheet) type)
+    nil))
+
+(defmethod request-selection-content :around (port name type)
+  (unless (or (typep type 'representation-type-name)
+              (and (listp type)
+                   (every (lambda (v) (typep v 'representation-type-name)) type)))
+    (error "Invalid type: ~s" type))
+  (call-next-method))
+
+(defgeneric request-clipboard-content (port sheet type)
+  (:documentation "Request the content of the clipboard.
+
+After calling this function, if the clipboard is active and can be
+presented as the requested type, an event of type CLIPBOARD-SEND-EVENT
+will be delivered to SHEET. This event will contain the requested data
+in one of the requested formats.
+
+TYPE is a list of keywords (or a single keyword, in which case it is
+interpreted as a list of a single element) that indicates the type of
+data that can be accepted. If the clipboard content can be represented
+in more than one format, the first matching type will be used.")
+  (:method ((port clim:port) (sheet clim:sheet) type)
+    nil))
+
+(defmethod request-clipboard-content :around (port name type)
+  (unless (every (lambda (v)
+                   (typep v 'representation-type-name))
+                 (alexandria:ensure-list type))
+    (error "Invalid type: ~s" type))
```

Comment:
also this contradicts that "unsupported types are ignored". it is not consistent.

Reply:
I wrote most of the details in the comment to the previous note. There was a bug which has not been fixed, but the main point is that the end-user API should check for calid types, but the backends should ignore types they are unable to process.